pkg/apis/k0s/v1beta1/clusterconfig_types.go

```diff
@@ -332,6 +336,64 @@ func (s *ClusterSpec) Validate() (errs []error) {
 		errs = append(errs, err)
 	}
 
+	for _, err := range s.validateKine(online) {
+		errs = append(errs, err)
+	}
+
+	return
+}
+
+func (s *ClusterSpec) validateKine(online bool) (errs field.ErrorList) {
+	if !online {
+		return
+	}
+
+	if s.Storage.Type != KineStorageType {
+		return
+	}
+
+	if s.Storage.Kine == nil {
+		errs = append(errs, field.NotFound(field.NewPath("storage", "kine"), "missing"))
+		return
+	}
+
+	dataSource := s.Storage.Kine.DataSource
+
+	if dataSource == "" {
+		// sqlite
+		return
+	}
+
+	path := field.NewPath("storage", "kine", "datasource")
+
+	u, err := url.Parse(dataSource)
+	if err != nil {
+		detail := "unable to parse datasource"
+		errs = append(errs, field.Invalid(path, dataSource, detail))
+		return
+	}
+
+	if u.Scheme == "sqlite" {
+		err = unix.Access(u.Path, unix.W_OK)
+		if err != nil {
+			detail := fmt.Sprintf("datasource (path) %s is not writable", dataSource)
+			errs = append(errs, field.Forbidden(path, detail))
+		}
+
+		return
+	}
+
+	// tcp service
+	conn, err := net.DialTimeout("tcp", net.JoinHostPort(u.Hostname(), u.Port()), time.Second)
```

Comment:
Not sure if timeout should be configurable, or maybe more "relaxed". I felt like 1 second should be enough. And this should work for all TCP services to see the the port can be connected to?

Reply:
Okay, to me this is a strong indicator that we indeed want to have the connectivity checks in a separate function. That way, we can pass in a context and rely on external cancellation/deadlines.

We can also use another error scheme then and not use the validation errors at all. Not sure if it makes sense to claim that a field value is invalid because a network peer is unreachable.